core/src/main/scala/kafka/server/RaftReplicaManager.scala

```diff
@@ -0,0 +1,397 @@
+
+            case HostedPartition.None =>
+              // Create the partition instance since it does not yet exist
+              (Some(Partition(topicPartition, time, configRepository, this)), None)
+          }
+          partition.foreach { partition =>
+            val isNew = priorDeferredMetadata match {
+              case Some(alreadyDeferred) => alreadyDeferred.isNew
+              case _ => prevPartitions.topicPartition(topicPartition.topic(), topicPartition.partition()).isEmpty
+            }
+            partitionChangesToBeDeferred.put(partition, isNew)
+          }
+        }
+
+        stateChangeLogger.info(s"Deferring metadata changes for ${partitionChangesToBeDeferred.size} partition(s)")
+        if (partitionChangesToBeDeferred.nonEmpty) {
+          delegate.makeDeferred(partitionChangesToBeDeferred, metadataOffset, onLeadershipChange)
+        }
+      } else { // not deferring changes, so make leaders/followers accordingly
+        val partitionsToBeLeader = mutable.HashMap[Partition, MetadataPartition]()
+        val partitionsToBeFollower = mutable.HashMap[Partition, MetadataPartition]()
+        builder.localChanged().foreach { currentState =>
+          val topicPartition = currentState.toTopicPartition
+          val partition = getPartition(topicPartition) match {
+            case HostedPartition.Offline =>
+              stateChangeLogger.warn(s"Ignoring handlePartitionChanges at $metadataOffset " +
+                s"for partition $topicPartition as the local replica for the partition is " +
+                "in an offline log directory")
+              None
+
+            case HostedPartition.Online(partition) => Some(partition)
+            case _: HostedPartition.Deferred => throw new IllegalStateException(
+              s"There should never be deferred partition metadata when we aren't deferring changes: $topicPartition")
+
+            case HostedPartition.None =>
+              // it's a partition that we don't know about yet, so create it and mark it online
+              val partition = Partition(topicPartition, time, configRepository, this)
+              allPartitions.putIfNotExists(topicPartition, HostedPartition.Online(partition))
+              Some(partition)
+          }
+          partition.foreach { partition =>
+            if (currentState.leaderId == localBrokerId) {
+              partitionsToBeLeader.put(partition, currentState)
+            } else {
+              partitionsToBeFollower.put(partition, currentState)
+            }
+          }
+        }
+
+        val prevPartitions = imageBuilder.prevImage.partitions
+        val changedPartitionsPreviouslyExisting = mutable.Set[MetadataPartition]()
+        builder.localChanged().foreach(metadataPartition =>
+          prevPartitions.topicPartition(metadataPartition.topicName, metadataPartition.partitionIndex).foreach(
+            changedPartitionsPreviouslyExisting.add))
+        val nextBrokers = imageBuilder.nextBrokers()
```

Comment:
rather than doing this, I would prefer to just give the ReplicaManager access to the BrokersBuilder and put the relevant accessors on that.  Otherwise, we create a bunch of objects that get thrown away (since nextBrokers will have to be called again later).  A small thing but it would be good to set up a good pattern here.

Reply:
> would prefer to just give the ReplicaManager access to the BrokersBuilder and put the relevant accessors on that. Otherwise, we create a bunch of objects that get thrown away (since nextBrokers will have to be called again later).

The point in time at which this invocation is occurring is when we are telling `RaftReplicaManager` to process the partition changes from a batch.  This happens after the batch has been otherwise handled by the metadata listener.  I think the most common occurrence for any batch is that there are no broker metadata changes (i.e. no broker has registered/unregistered or fenced/unfenced), in which case invoking `nextBrokers()` simply returns the previous brokers -- we don't create any objects.  If we make the change I think you are suggesting and `RaftReplicaManager` were to invoke `imageBuilder.brokersBuilder().build()` then we would actually invoke `MetadataBrokers(log, newBrokerMap)`, which is much more expensive.  Maybe you are suggesting `MetadataBrokersBuilder` should remember that it was already built?  It sort of does in that any attempt to mutate it after it is built will fail (since its map is now unmodifiable), but calling `build()` a second time could be made more efficient than it is now.

Reply:
Thanks for the explanation!

Let's just rename this to "brokers" and then I think we're good.  (Everything in the builder is "next" after all :) )

Reply:
As discussed offline, will keep it like it is but rename `nextBrokers()` to simply `brokers()`.